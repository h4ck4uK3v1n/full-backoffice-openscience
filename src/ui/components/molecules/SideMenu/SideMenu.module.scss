@import '../../../../globals/ads-variables';
@import '../../../../globals/mixins';
@import '../../../../globals/_fonts';

.side-menu {
  background-color: $--ads-colors-black;
  color: $--ads-colors-white;
  display: flex;
  flex-direction: column;
  font-family: Barlow, sans-serif;
  transition:
    width 0.5s ease,
    padding 0.5s ease;
  width: 250px;

  @include respond-to('small-mobile') {
    width: 200px;
  }

  @include respond-to('medium') {
    width: 220px;
  }

  @include respond-to('large') {
    width: 270px;
  }

  @include respond-to('extra-large') {
    width: 300px;
  }

  &.collapsed {
    width: 60px;

    @include respond-to('small-mobile') {
      width: 50px;
    }

    @include respond-to('medium') {
      width: 55px;
    }

    @include respond-to('large') {
      width: 65px;
    }

    @include respond-to('extra-large') {
      width: 70px;
    }
  }
}

.menu-header {
  align-items: center;
  display: flex;
  height: 60px;
  justify-content: center;
  margin-right: 10px;
  transition: all 0.3s ease;

  .hide-menu-button {
    align-items: center;
    background: none;
    border: none;
    color: $--ads-colors-white;
    cursor: pointer;
    display: flex;
    font-family: Barlow, sans-serif;
    font-size: 20px;
    gap: 10px;
    text-transform: uppercase;
    transition:
      justify-content 0.3s ease,
      gap 0.3s ease;

    .icon {
      align-items: center;
      display: flex;
      height: 20px;
      justify-content: center;
      width: 20px;
    }

    span {
      display: none;
      transition: opacity 0.3s ease;
    }

    &:not(.collapsed) span {
      display: inline;
      opacity: 1;
    }
  }
}

.menu-list {
  display: flex;
  flex-direction: column;
  padding: 0;

  .side-menu-button {
    align-items: center;
    background-color: $--ads-colors-black;
    border: none;
    color: $--ads-colors-white;
    cursor: pointer;
    display: flex;
    font-family: Barlow, sans-serif;
    font-size: 16px;
    height: 50px;
    justify-content: flex-start;
    letter-spacing: 0.16px;
    line-height: 30px;
    padding: 10px 15px;
    text-align: left;
    transition: all 0.3s ease;

    svg {
      fill: $--ads-colors-white;
      height: 24px;
      margin-right: 10px;
      transition: margin 0.3s ease;
      width: 24px;
    }

    &:hover {
      background-color: $--ads-colors-gray-20;
      color: $--ads-colors-gray-70;

      svg {
        fill: $--ads-colors-gray-70;
      }
    }

    &.collapsed {
      justify-content: center;
      padding: 0;
      text-align: center;

      svg {
        margin-right: 0;
      }

      span {
        display: none;
      }
    }
  }
}
