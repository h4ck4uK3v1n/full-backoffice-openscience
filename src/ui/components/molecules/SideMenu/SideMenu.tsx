import React, { useState } from 'react';
import type { ISideMenuProps } from './types/Iprops';
import SideMenuButton from '../../atoms/side-menu-button/SideMenuButton';
import styles from './SideMenu.module.scss';
import circleLeftIcon from '../../../../assets/icons/circle-left.svg?raw';
import linesIcon from '../../../../assets/icons/lines.svg?raw';

function SideMenu({ menuItems, onClick }: ISideMenuProps) {
  const [isCollapsed, setIsCollapsed] = useState(false);

  const toggleMenu = () => {
    setIsCollapsed((prev) => !prev);
  };

  return (
    <aside
      data-testid="side-menu"
      className={`${styles['side-menu']} ${isCollapsed ? styles.collapsed : ''}`}
    >
      <div className={styles['menu-header']}>
        <button className={styles['hide-menu-button']} onClick={toggleMenu}>
          <span
            className={styles.icon}
            dangerouslySetInnerHTML={{
              __html: isCollapsed ? linesIcon : circleLeftIcon,
            }}
          />
          {!isCollapsed && 'HIDE MENU'}
        </button>
      </div>
      <nav className={styles['menu-list']}>
        {menuItems.map((item) => (
          <SideMenuButton
            key={item.id}
            icon={item.icon}
            path={item.path}
            basePath={item.basePath}
            className={isCollapsed ? styles.collapsed : ''}
            onClick={onClick}
          >
            {isCollapsed ? null : item.title}
          </SideMenuButton>
        ))}
      </nav>
    </aside>
  );
}

export default SideMenu;
