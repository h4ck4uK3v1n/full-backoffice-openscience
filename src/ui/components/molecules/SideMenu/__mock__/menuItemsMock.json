{
    "data": [
      {
        "id": 1,
        "title": "Content Management",
        "icon": "content-management",
        "path": "/article",
        "basePath": "/"
      },
      {
        "id": 2,
        "title": "Article Management",
        "icon": "article-management",
        "path": "/author",
        "basePath": "/"
      },
      {
        "id": 3,
        "title": "Author Management",
        "icon": "author-management",
        "path": "/bibliographies",
        "basePath": "/"
      },
      {
        "id": 4,
        "title": "Bibliographines",
        "icon": "bibliographines",
        "path": "/gallery",
        "basePath": "/"
      },
      {
        "id": 5,
        "title": "Image Gallery",
        "icon": "image-gallery",
        "path": "/index",
        "basePath": "/"
      },
      {
        "id": 6,
        "title": "Index Builder",
        "icon": "index-builder",
        "path": "/volume",
        "basePath": "/"
      },
      {
        "id": 7,
        "title": "Volume Management",
        "icon": "volume-management",
        "path": "/volume",
        "basePath": "/"
      },
      {
        "id": 8,
        "title": "Site Configuration",
        "icon": "site-configuration",
        "path": "/volume",
        "basePath": "/"
      },
      {
        "id": 9,
        "title": "SEO Settings",
        "icon": "seo-settings",
        "path": "/volume",
        "basePath": "/"
      },
      {
        "id": 10,
        "title": "Dynamic Banner Management",
        "icon": "dinamicBannerManagement",
        "path": "/volume",
        "basePath": "/"
      },
      {
        "id": 11,
        "title": "User Management",
        "icon": "user-management",
        "path": "/volume",
        "basePath": "/"
      }
    ]
  }
  