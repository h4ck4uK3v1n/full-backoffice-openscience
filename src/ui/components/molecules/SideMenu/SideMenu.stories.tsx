import type { StoryObj, Meta } from '@storybook/react';
import SideMenu from './SideMenu';

const meta: Meta<typeof SideMenu> = {
  title: 'ui/components/molecules/SideMenu',
  component: SideMenu,
  argTypes: {
    onClick: {
      action: 'click',
    },
  },
};

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    menuItems: [
      {
        id: '1',
        title: 'Content Management',
        icon: 'content-management',
        path: '/Content-Management',
        basePath: '/',
      },
      {
        id: '2',
        title: 'Article Management',
        icon: 'article-management',
        path: '/Article-Management',
        basePath: '/',
      },
      {
        id: '3',
        title: 'Author Management',
        icon: 'author-management',
        path: '/Author-Management',
        basePath: '/',
      },
      {
        id: '4',
        title: 'Bibliographines',
        icon: 'bibliographines',
        path: '/Bibliographines',
        basePath: '/',
      },
      {
        id: '5',
        title: 'Image Galley',
        icon: 'image-gallery',
        path: '/Image-Galley',
        basePath: '/',
      },
      {
        id: '6',
        title: 'Index Builder',
        icon: 'index-builder',
        path: '/Index-Builder',
        basePath: '/',
      },
      {
        id: '7',
        title: 'Volume Management',
        icon: 'volume-management',
        path: '/Volume-Management',
        basePath: '/',
      },
      {
        id: '8',
        title: 'Site Configuration',
        icon: 'site-configuration',
        path: '/Site-Configuration',
        basePath: '/',
      },
      {
        id: '9',
        title: 'SEO Settings',
        icon: 'seo-settings',
        path: '/SEO-Settings',
        basePath: '/',
      },
      {
        id: '10',
        title: 'Dynamic Banner Management',
        icon: 'dinamicBannerManagement',
        path: '/Dynamic-Banner-Management',
        basePath: '/',
      },
      {
        id: '11',
        title: 'User Management',
        icon: 'user-management',
        path: '/User-Management',
        basePath: '/',
      },
    ],
  },
};
