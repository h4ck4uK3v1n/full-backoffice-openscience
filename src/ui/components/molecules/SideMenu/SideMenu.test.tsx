import { describe, it, expect, vi } from 'vitest';
import { render, screen, fireEvent } from '@testing-library/react';
import SideMenu from './SideMenu';
import type { ISideMenuProps } from './types/Iprops';
import styles from './SideMenu.module.scss';
import menuItemsMock from './__mock__/menuItemsMock.json';

vi.mock('../../../../assets/icons/content-management.svg?raw', () => ({
  default: 'test-file-stub',
}));
vi.mock('../../../../assets/icons/article-management.svg?raw', () => ({
  default: 'test-file-stub',
}));
vi.mock('../../../../assets/icons/author-management.svg?raw', () => ({
  default: 'test-file-stub',
}));
vi.mock('../../../../assets/icons/bibliographines.svg?raw', () => ({ default: 'test-file-stub' }));
vi.mock('../../../../assets/icons/image-gallery.svg?raw', () => ({ default: 'test-file-stub' }));
vi.mock('../../../../assets/icons/index-builder.svg?raw', () => ({ default: 'test-file-stub' }));
vi.mock('../../../../assets/icons/volume-management.svg?raw', () => ({
  default: 'test-file-stub',
}));
vi.mock('../../../../assets/icons/site-configuration.svg?raw', () => ({
  default: 'test-file-stub',
}));
vi.mock('../../../../assets/icons/seo-settings.svg?raw', () => ({ default: 'test-file-stub' }));
vi.mock('../../../../assets/icons/dinamicBannerManagement.svg?raw', () => ({
  default: 'test-file-stub',
}));
vi.mock('../../../../assets/icons/user-management.svg?raw', () => ({ default: 'test-file-stub' }));

const mockMenuItems: ISideMenuProps['menuItems'] = menuItemsMock.data;

const defaultProps: ISideMenuProps = {
  menuItems: mockMenuItems,
  onClick: vi.fn(),
};

describe('Testing for SideMenu component', () => {
  it('renders the SideMenu component', () => {
    render(<SideMenu menuItems={defaultProps.menuItems} onClick={defaultProps.onClick} />);
    expect(screen.getByText('HIDE MENU')).toBeTruthy();
  });

  it('toggles menu collapse state', () => {
    render(<SideMenu menuItems={defaultProps.menuItems} onClick={defaultProps.onClick} />);
    const button = screen.getByText('HIDE MENU');
    fireEvent.click(button);
    expect(screen.queryByText('HIDE MENU')).toBeNull();
    const menu = screen.getByTestId('side-menu');
    expect(menu.classList.contains(styles.collapsed)).toBe(true);
  });

  it('renders menu items', () => {
    render(<SideMenu menuItems={defaultProps.menuItems} onClick={defaultProps.onClick} />);
    mockMenuItems.forEach((item) => {
      expect(screen.getByText(item.title)).toBeTruthy();
    });
  });
});
