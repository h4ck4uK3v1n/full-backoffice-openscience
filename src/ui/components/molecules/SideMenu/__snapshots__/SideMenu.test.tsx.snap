// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SideMenu Snapshot > renders correctly 1`] = `
<DocumentFragment>
  <aside
    class="_side-menu_54da13"
    data-testid="side-menu"
  >
    <div
      class="_menu-header_54da13"
    >
      <button
        class="_hide-menu-button_54da13"
      >
        <span
          class="_icon_54da13"
          dangerouslySetInnerHTML="[object Object]"
        />
        HIDE MENU
      </button>
    </div>
    <nav
      class="_menu-list_54da13"
    >
      <button
        class="_side-menu-button_bf946c _side-menu-button--primary_bf946c"
        data-testid="side-menu-button-styles"
      >
        <span
          class="_icon_bf946c"
          style="width: 24px; height: 24px; display: block;"
        />
        Content Management
      </button>
      <button
        class="_side-menu-button_bf946c _side-menu-button--primary_bf946c"
        data-testid="side-menu-button-styles"
      >
        <span
          class="_icon_bf946c"
          style="width: 24px; height: 24px; display: block;"
        />
        Article Management
      </button>
      <!-- Más elementos de menú aquí -->
    </nav>
  </aside>
</DocumentFragment>
`;
